### PR TITLE
Improve DatasetOptimizer API

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR improves the API of the dataset optimizer and add the API over the DataModule.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18827.org.readthedocs.build/en/18827/

cc @borda

### CODE REVIEW

File: src/lightning/app/cli/commands/cp.py [hunk 1]
```diff
@@ -131,7 +131,7 @@ def _upload_files(live, client: LightningClient, local_src: str, remote_dst: str
             )
             upload_urls.append(response)
 
-    upload_urls = [upload_url.get().upload_url for upload_url in upload_urls]
+    upload_urls = [upload_url.get().upload_id for upload_url in upload_urls]
```
Comment (justusschock) [inline src/lightning/app/cli/commands/cp.py:134]: this wont work with the new API, you'll need to extend the list by `[url.url for url in upload_url.get().urls]` also make a call to to complete the upload correctly

File: tests/tests_app/cli/test_cp.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def test_cp_local_to_remote(tmpdir, monkeypatch):
     )
 
     result = MagicMock()
-    result.get.return_value = V1UploadProjectArtifactResponse(upload_url="http://foo.bar")
+    result.get.return_value = V1UploadProjectArtifactResponse(upload_id="http://foo.bar")
```
Comment (justusschock) [inline tests/tests_app/cli/test_cp.py:50]: see my comment above for why this test doesn't test the reality

**Review by justusschock — CHANGES_REQUESTED** [review]
Data API looks good. CLI fix won't work

File: tests/tests_app/cli/test_cp.py (hunk 2 above)
Comment (tchaton) [reply tests/tests_app/cli/test_cp.py]: done .

**Review by ethanwharris — APPROVED** [review]

File: src/lightning/data/streaming/dataset.py [hunk 3]
```diff
@@ -11,36 +11,135 @@
...
+from torch.utils.data import IterableDataset
 
+from lightning.data.datasets.env import _DistributedEnv, _WorkerEnv
 from lightning.data.streaming import Cache
+from lightning.data.streaming.item_loader import BaseItemLoader
+from lightning.data.streaming.sampler import ChunkedIndex
 
 
-class StreamingDataset(Dataset):
+class StreamingDataset(IterableDataset):
```
Comment (lantiga) [inline src/lightning/data/streaming/dataset.py:25]: Just a check if I understand well: the change in the base class is for the dataloader to consume this as an iterable dataset (using `__iter__`), but `__len__` and `__getitem__` are also implemented. These are convenience methods for users at this point, since the dataloader will not use them. Which also means a custom Sampler won't be used.

But at the same time, `__next__` calls them, so it's a mapped dataset with an iterable dataset facade. What was the rationale for this change?

File: src/lightning/data/streaming/dataset.py [hunk 4]
```diff
@@ -11,36 +11,135 @@
...
+    def __next__(self) -> Any:
+        # Lazily re-populate the interval to reduce memory usage.
+        if len(self.current_indexes) == 0:
+            if self.chunk_index == len(self.worker_intervals):
+                raise StopIteration
+
+            interval = self.worker_intervals[self.chunk_index]
+            current_indexes = np.arange(0, interval[1] - interval[0])
+            if self.shuffle:
+                current_indexes = self.random_state.permutation(current_indexes)
```
Comment (lantiga) [inline src/lightning/data/streaming/dataset.py:126]: do we need to determine how we permute the indices ahead of time to make this fault tolerant? Or we just keep the `random_state` around to eventually replay back?

File: src/lightning/data/streaming/dataset_optimizer.py [hunk 5]
```diff
@@ -547,9 +510,27 @@ def __init__(
...
+    def run(self, optimizable_dataset: _OptimizableDataset) -> None:
         """The `DatasetChunker.run(...)` method is used to trigger the data processing from your dataset into
         chunks."""
+        if not isinstance(optimizable_dataset, _OptimizableDataset):
+            raise ValueError(
+                """The provided argument to the DatasetOptimizer.run(...) needs to have the following format:
+
+                Example:
+
+                    class YourDataset:
```
Comment (lantiga) [inline src/lightning/data/streaming/dataset_optimizer.py:522]: Relying on protocols feels very different from the rest of Lightning.
Shouldn't we just rely on providing a class with clear hooks that the user needs to implement?

File: src/lightning/app/cli/commands/cp.py [hunk 6]
```diff
@@ -129,9 +129,11 @@ def _upload_files(live, client: LightningClient, local_src: str, remote_dst: str
                 body=ProjectIdStorageBody(cluster_id=cluster.cluster_id, filename=filename),
                 async_req=True,
             )
-            upload_urls.append(response)
+            _upload_urls.append(response)
 
-    upload_urls = [upload_url.get().upload_url for upload_url in upload_urls]
+    upload_urls = []
+    for upload_url in _upload_urls:
+        upload_urls.append(upload_url.get().urls)
```
Comment (justusschock) [inline src/lightning/app/cli/commands/cp.py:136]:
```suggestion
        upload_urls.extend(upload_url.get().urls)
```

**Review by justusschock — APPROVED** [review]

**Review by lantiga — COMMENTED** [review]
Looks good but I'd like to get a couple of clarifications (in the comments)

File: src/lightning/data/streaming/dataset.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:25]: Users don't know about IterableDataset and they don't need to. How we handle the sampling under the hood is up to us.
However, from the user perceptive, this behaves exactly as a map based dataset with a simple API. As a user, I just need to override the `__getitem__` method and voila.

```python
class ImageNetStreamingDataset(StreamingDataset):

    def __init__(self, name, version):
        super().__init__(name=name, version=version)
        self.transform = T.Compose([
            T.RandomResizedCrop((224, 224), antialias=True),
        ])

    def __getitem__(self, index):
        # Note: If torchvision is installed, we return a tensor image instead of a pil image as it is much faster. 
        img, class_index = super().__getitem__(index) # <- Whatever you returned from the DatasetOptimizer prepare_item method.
        return self.transform(to_rgb(img) / 255.), int(class_index)  
```

File: src/lightning/data/streaming/dataset.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:126]: Yes, this is left for another PR.

File: src/lightning/data/streaming/dataset_optimizer.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset_optimizer.py:522]: We can once we stabilise the API. @carmocca was against having this on the DataModule even If I believe this provides an easy onboarding.

Ideally, I would like to avoid adding a new object the user has to learn about just for 2 hooks. This seems like an overkill.

File: src/lightning/data/streaming/dataset.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/data/streaming/dataset.py:25]: This change has some implications for distributed sampling, because now the responsibility of sampling is with the dataset implementation, it also needs to include the dataloader-workers in the calculation, otherwise we end up with data duplication. Can we add a test that the dataset behaves correctly and equivalently to a regular dataset/dataloader with distributed sampler + multiple dataloader workers?

File: src/lightning/data/streaming/dataset.py [hunk 7]
```diff
@@ -11,36 +11,135 @@
...
+        self.min_items_per_replica: Optional[int] = None
+        self.seed = seed
+        self.num_iter = 0
+        self.random_state = None
 
     def __len__(self) -> int:
-        return len(self.cache)
+        return self.L
+
+    def __iter__(self) -> "StreamingDataset":
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:77]: The lines below are important to cover with a test I think

Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:77]: Yes, this is coming. This isn't supported yet for distributed training.

File: src/lightning/data/streaming/dataset.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:25]: Yes. But this is something we handle. So we just need to make this solid.

File: src/lightning/data/streaming/dataset_optimizer.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset_optimizer.py:524]: Yes, that's entirely fine. Let's get user feedbacks first.

Comment (tchaton) [reply src/lightning/data/streaming/dataset_optimizer.py:524]: Resolving for now as it would be addressed promptly.

**Review by lantiga — APPROVED** [review]
Approving for expediency, the API will need revision
